### PR TITLE
Features: Private-events

### PR DESCRIPTION
**Pull request by kirokoNjenga**
-Users can create events and send invitations and parties. 
-Events take place at a specific date and at a location. 
-An application named Private Events that behaves similar to Eventbrite. 
-The main goal is to put in practice the main concepts of Associations in rails.

### CODE REVIEW

File: app/views/layouts/_header.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  <%= link_to "private events app", root_path, id: "logo" %>
+  <ul class="nav navbar-nav navbar-right">
+    <% if signed? %>
+      <li class="user-btn "><%= link_to "#{current_user.name.split(" ")[0]} Events" , user_path(current_user) %></li>
+      <li><%= link_to "New Event", new_event_path %></li>
+      <li><%= link_to "Logout", logout_path %></li>
+    <% else %>
+      <li><%= link_to "Sign up", new_user_path %></li>
+      <li><%= link_to "Sign in", signin_path %></li>
+    <% end %>
```
Comment (euqueme) [inline app/views/layouts/_header.html.erb]: * [ ] There should be absolutely NO logic in views. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

File: app/views/shared/_error_messages.html.erb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    <div class="alert alert-error">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
+    </div>
+    <ul>
+      <% object.errors.full_messages.each do |msg| %>
+        <li>* <%= msg %></li>
+      <% end %>
+    </ul>
+  </div>
+<% end %>
```
Comment (euqueme) [inline app/views/shared/_error_messages.html.erb]: * [ ] There should be absolutely NO logic in views. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Samkiroko :balloon: 

### Thank you for your submission :medal_sports: your project has a great start, I personally like the fact that you can choose many attendees when creating a new event :star: There are still some changes to make :construction: :construction_worker: 

#### Behaviour

- [x] Event's index page displays past and upcoming events lists ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- I had to put the URL in the browser to see this page, this should be the root page so please fix it in your routes
image

- [x] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- Instead of "Past events created by User:" The title of this section should be "Past events booked to attend to by User:"
image

- [x] Navigation links are present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- When you click in PRIVATE EVENTS APP at the left corner of the page it should take you to the events index page

#### Views and Helpers
* [x] There should be absolutely NO logic in views. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**

* You have a couple if/end and if/else statements check the following material that guides you to using helpers for if statements [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

#### General

- [x] Files not related to the project should not be committed and added to .gitignore (e.g. `.DS_Store` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

#### Tests

* [x] models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

- In addition, Rspec, and Capybara doesn't seem to be configured and your app is throwing an error when trying to execute your test cases

image

### That's it! :raised_hands: 

### Please :pray: do the changes :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Comment by kirokoNjenga** [comment]
**Requested changes by TSE**
* Changed the routes to display all listed events instead of the home page.
* The users show page displays (click {your name Events} if logged_in): 
- [ ] The event that the user has created.                                                       
- [ ] The event that the user is attending.
* Changed Instead of "Past events created by User:" to "Past events booked to attend to by User:"
* change  PRIVATE EVENTS APP logo at the left corner of the page to redirect to the events index page.
*  There should be absolutely NO logic in views the two files: `app/views/layouts/_header.html.erb`
`app/views/shared/_error_messages.html.erb`
- I chose the Partials instead of the helper because they are reusable across different views of the application as the best practice.
* Added `.DS_Store` to .inginore file
* Rspec, and Capybara has been configured in the app:
to run the test cases run the following commands

-Model test cases:

`$ rspec spec/models/`

-Integration test cases:

`$ rspec spec/integration/`

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @Samkiroko, 👋 

### Great job implementing some of the suggested changes. 🙌 Please see my comments. 👇 

- [x] Helpers can also be used across the app, you just have to add it to the `application_helper.rb`. Besides, having logic in the views is actually a bad practice, regardless of whether it's a partial or normal view. Views are views. Check the last point [here](https://dzone.com/articles/9-best-practices-to-follow-while-coding-in-rails-1) for more clarification. You could check other bad practices to avoid there too.

- [x] Fix your linter errors.

- [x] Fix the merge conflicts.

**Kindly submit another CRR once these issues have been resolved.** _Happy coding!_ 🎉

**Comment by kirokoNjenga** [comment]
* Removed logic from view to helper
* cleared all errors and marge conflict

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Samkiroko :handshake: 

### Thank you for doing the changes requested by the previous reviewer :medal_sports: your project is working better and better every time. there are still some things to work on from my first review and a couple of things breaking so let's fix them :construction: :construction_worker: 

#### Behaviour

- [x] Your app is throwing an error in the sign in.

image

This is because you removed the if/else statement to render the links but didn't create helper methods to handle the logic. Next, you'll find the code to properly use helper and partials to remove logic from your views:

First, you needed to create a method in your SessionsHelper that renders one partial for the links in the if and another for the links in the else statement

app/helpers/sessions_helper.rb
```ruby
  def render_links
    if signed?
      render partial: 'layouts/links_signed'
    else
      render partial: 'layouts/links_not_signed'
    end
  end
```
then you have to create both partials
app/views/layouts/_links_signed.html.erb
```html
<li class="user-btn "><%= link_to "#{current_user.name.split(" ")[0]} Events" , user_path(current_user) %></li>
<li><%= link_to "New Event", new_event_path %></li>
<li><%= link_to "Logout", logout_path %></li>
```
app/views/layouts/_links_not_signed.html.erb
```html
<li><%= link_to "Sign up", new_user_path %></li>
<li><%= link_to "Sign in", signin_path %></li>
```
Finally, you have to call the method created in the SessionsHelper within the _header.html.erb file
```html
<header class="navbar navbar-fixed-top navbar-custom">
  <%= link_to "private events app", events_path, id: "logo" %>
  <ul class="nav navbar-nav navbar-right">
    <%= render_links %>
  </ul>
</header>
```
#### Test cases

- Your project still doesn't cover the test cases requested in my first review and is still throwing an error when trying to execute them, so please properly configure RSpec and capybara a build test cases for your project
image


### That's it! :raised_hands: 

### Please :pray: do the changes :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES  :traffic_light:

**Comment by kirokoNjenga** [comment]
* Thank you so much, the helper function worked
image
*Also the unite test is working well
image

**Comment by uryelah** [comment]
Hi **Samuel Njenga**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job going through the review process and resolving all the pointed errors.

This project is approved, but just remember to add the DS_store and other unrelated files to the gitigore before merging and in the next projects.

- Files not related to the project should not be committed and added to .gitignore

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**

---------

    
    
**Keep being an awesome dev and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Pull request merged**